internal/controller/imageprefetch_controller.go

```diff
@@ -2,45 +2,508 @@ package controller
+			Type:    ofenv1.ConditionImagePullFailed,
+			Status:  metav1.ConditionTrue,
+			Reason:  "ImagePrefetchFailed",
+			Message: "some nodes failed to pull the image",
+		})
+	} else {
+		meta.SetStatusCondition(&imgPrefetch.Status.Conditions, metav1.Condition{
+			Type:    ofenv1.ConditionProgressing,
+			Status:  metav1.ConditionTrue,
+			Reason:  "ImagePrefetchProgressing",
+			Message: "Waiting for all nodes to pull the image",
+		})
+	}
+
+	return result, r.Status().Update(ctx, imgPrefetch)
+}
+
+type NodeImageSetStatus struct {
+	desiredNodes    int
+	availableNodes  int
+	pullingNodes    int
+	pullFailedNodes int
+}
+
+func calculateStatus(selectNodes []string, nodeImageSets *ofenv1.NodeImageSetList) NodeImageSetStatus {
+	status := NodeImageSetStatus{}
+	status.desiredNodes = len(selectNodes)
+
+	for _, nodeImageSet := range nodeImageSets.Items {
+		if meta.IsStatusConditionTrue(nodeImageSet.Status.Conditions, ofenv1.ConditionImageAvailable) {
+			status.availableNodes++
+		}
+		if meta.IsStatusConditionTrue(nodeImageSet.Status.Conditions, ofenv1.ConditionImageDownloadFailed) {
+			status.pullFailedNodes++
+		}
+		if meta.IsStatusConditionTrue(nodeImageSet.Status.Conditions, ofenv1.ConditionImageDownloadComplete) &&
+			!meta.IsStatusConditionTrue(nodeImageSet.Status.Conditions, ofenv1.ConditionImageAvailable) {
+			status.pullingNodes++
+		}
+	}
+
+	return status
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ImagePrefetchReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	nodeImageSetHandler := handler.EnqueueRequestsFromMapFunc(
+		func(ctx context.Context, obj client.Object) []ctrl.Request {
+			nodeImageSet := obj.(*ofenv1.NodeImageSet)
+			imagePrefetch := &ofenv1.ImagePrefetch{}
+			err := r.Get(ctx, types.NamespacedName{
+				Namespace: nodeImageSet.Labels[constants.OwnerImagePrefetchNamespace],
+				Name:      nodeImageSet.Labels[constants.OwnerImagePrefetchName],
+			}, imagePrefetch)
+			if err != nil {
+				return nil
+			}
+
+			return []ctrl.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Namespace: imagePrefetch.Namespace,
+						Name:      imagePrefetch.Name,
+					},
+				},
+			}
+		})
+
+	nodeHandler := handler.EnqueueRequestsFromMapFunc(
+		func(ctx context.Context, obj client.Object) []ctrl.Request {
+			node := obj.(*corev1.Node)
+			imagePrefetchList := &ofenv1.ImagePrefetchList{}
+			err := r.List(ctx, imagePrefetchList)
+			if err != nil {
+				return nil
+			}
+
+			var requests []ctrl.Request
+			for _, imgPrefetch := range imagePrefetchList.Items {
+				if imgPrefetch.Spec.NodeSelector == nil {
+					continue
+				}
+
+				for key, value := range imgPrefetch.Spec.NodeSelector {
+					if node.Labels[key] == value {
+						requests = append(requests, ctrl.Request{
+							NamespacedName: types.NamespacedName{
+								Namespace: imgPrefetch.Namespace,
+								Name:      imgPrefetch.Name,
+							},
+						})
+					}
+				}
+			}
```

Comment:
Can this logic make proper requests?
If the `.spec.nodeSelector` is not defined in ImagePrefetch CR, the reconciliation for the CR will not be run when node is deleted.
Should we check the `.status.selectedNodes`?

Reply:
I intended to trigger the Reconcile of the ImagePrefetch controller by having the NodeImageSet controller delete the NodeImageSet when a Node resource is deleted. I modified the implementation to use .status.selectedNodes instead, as it makes the mechanism simpler.
https://github.com/cybozu-go/ofen/pull/8/commits/78092285bdf8934fc97c31746b0ea8478b3002ae